### DISCUSSION TITLE
Matrix multiplication as summation

### DISCUSSION THREAD

**Discussion by ArjunM21** [yalmip/YALMIP#1044] [General discussion]
Hi all.
I had previously asked a question regarding YALMIP that I finally got a good answer and was able to solve it as I required. Now I am stuck again. I am trying to code a specific problem, given in the attached PDF file. I am trying the equation 4 as it is this time. Please find my code below.

`clc;
clear all;
close all;
% C = [1 0 0 ; 1 1 -1; 1 1 -1; 1 1 -1; 0 1 0; 1 0 -1; 0 1 -1; 0 0 1]; %1,5,8
C = [1 0 0 ; 0 1 0; 0 1 0; 0 1 0; 0 0 1; 0 1 -1;-1 1 0; 1 -1 1]; %1,2,5

sdpvar t;
q = binvar(8,1);
Q = diag(q);
q0 = [0;0;0;0;0;1;1;1];
Q0 = diag(q0);

syms W W1 W2 W3 W4 W5 W6 W7 W8
W = [W1 W2 W3 W4 W5 W6 W7 W8];
W1 = intvar(3,3);
W2 = intvar(3,3);
W3 = intvar(3,3);
W4 = intvar(3,3);
W5 = intvar(3,3);
W6 = intvar(3,3);
W7 = intvar(3,3);
W8 = intvar(3,3);
a = sum(W*q);
eq1 = sum(W*q) == C' * Q * C;
eq2 = sum(W) == C' * C;

Model = [];
Model = [sum(q)==3]; 
ch = 100;
tol = 1.4e1;

Gdif = zeros(1,8);
for i = 1:8
    Gdif(:,i) = -trace(inv(a) * value(W(i)) * inv(a) * (C'*C));;
end

% while(ch > tol)
%     G0 = trace((C/(C'*Q0*C))*C');
%     Model = [Model, G0 + Gdif * (q-q0) <= t];
%     ss = optimize(Model,t);
%     q0 = round(value(q));
%     Q0 = diag(q0);
%     Qt = sum(q0);
%     t0 = round(value(t));
%     display(trace(C/(C'*Q0*C)*C'));
%     ch = abs(t0 - trace(C/(C'*Q0*C)*C'));
end `

Here, C is a pre-defined matrix. I need to express C' * Q * C as the sum of W * q. Is there a good way to do this? I am getting error in the step for Gdif as I do not know how to convert sym to a double . I tried using double directly, but its an error. 

Also, is there a way for me to see the actual variables that are being chosen on case of q? I mean like q1, q2 etc in the command window? I tried typing q in the command window, but just got the information regarding q, not the actual variable names. Disp(q) also did not give me anything good.

And, is there another way to define my W vector containing all other Ws? My code looks off. All the commented section is to be changed accordingly as well.

Thank you in advance.
[Optimal sensor placement strategies for large scale.pdf](https://github.com/yalmip/YALMIP/files/7602392/Optimal.sensor.placement.strategies.for.large.scale.pdf)

**Comment by johanlofberg** [comment]
Very unclear what you want to do. C' * Q * C is a matrix, while W * q is a vector, so they are not comparable. If you for some reason want to extract the low-level linear basis for vec( C' * Q * C)

```
>> C = randn(2);
>> q = sdpvar(2,1);Q = diag(q);
>> S = C'*Q*C;
>> W = getbase(S);
>> W*[1;q] - S(:)

ans =

     0
     0
     0
     0

>> reshape(W*[1;q],2,2) - S

ans =

     0     0
     0     0

```


...and please put your code in code blocks <>

**Reply by johanlofberg** [reply]
```
>> yalmip('clear');
>> q = sdpvar(2,1);
>> q(1)
Linear scalar (real, 1 variable)
Coefficients range: 1 to 1
>> sdisplay(q(1))
internal(1)
>> sdisplay(q)

ans =

  2×1 cell array

    {'internal(1)'}
    {'internal(2)'}

>> sdisplay(recover(1))
internal(1)

>> recover(1)-q(1)

ans =

     0


```

**Comment by ArjunM21** [comment]
Hey thank you for your suggestions. 
First of all, W is a vector of different 3x3 matrices W1, W2, W3 etc. q is a vector with q1, q2, q3 etc which are binary variables. C'*Q*C is a 3x3 matrix, while I think W*q would give something like W1q1, W2q2 etc, where each W is a 3x3 matrix and q has either a value of 0 or 1. In that case, C'*Q*C and W*q can be compared right?

Also, when I checked the sdisplay function, my q1 shows up as internal 2 i.e. q values are from internal 2 to internal 9 instead of 1 to 8. Is there a reason for this?

Let me explain what I want with this problem. I had already asked a question regarding this paper previously [https://github.com/yalmip/YALMIP/discussions/1028](url). In this I tried to solve the problem using C'*Q*C alone. Here I am going to replace C'*Q*C with sigma(Wiqi) and trying to code everything as in the paper or the link provided above.

**Reply by johanlofberg** [reply]
internal(x) simply means that variable was the x:th variable defined internally since matlab started or last time yalmip('clear') was run, and is used as a name when YALMIP cannot figure out what the variable is called in the work-space.

I think you are on your own here, if you really need to dig into low-level formats to get the basis and internal variables and what not, you simply have to dig into that
https://yalmip.github.io/command/recover/
https://yalmip.github.io/command/getvariables/
https://yalmip.github.io/command/getbase/

**Comment by ArjunM21** [comment]
Okay thank you for your help.

I have kind of done the coding part, but gets an error saying "The first argument should be an sdpvar object or string." for the line after calculating G0. Can you help me there? Pls find the code below

`clc;
clear all;
close all;
% C = [1 0 0 ; 1 1 -1; 1 1 -1; 1 1 -1; 0 1 0; 1 0 -1; 0 1 -1; 0 0 1]; %1,5,8
C = [1 0 0 ; 0 1 0; 0 1 0; 0 1 0; 0 0 1; 0 1 -1;-1 1 0; 1 -1 1]; %1,2,5

sdpvar t;
q = binvar(8,1);
Q = diag(q);
q0 = [0;0;0;0;0;1;1;1];
Q0 = diag(q0);

syms W W1 W2 W3 W4 W5 W6 W7 W8
W = [W1 W2 W3 W4 W5 W6 W7 W8];
W1 = intvar(3,3);
W2 = intvar(3,3);
W3 = intvar(3,3);
W4 = intvar(3,3);
W5 = intvar(3,3);
W6 = intvar(3,3);
W7 = intvar(3,3);
W8 = intvar(3,3);
a = sum(W*q);
eq1 = sum(W*q) == C'*Q*C;
eq2 = sum(W) == C'*C;

Model = [];
Model = [sum(q) == 3]; 
ch = 100;
tol = 1.4e1;

Gdif = [];
for i = 1:8
    Gdif = [Gdif, -trace(inv(a)*value(W(i))*inv(a)*(C'*C))];
end

while(ch > tol)
    G0 = trace((C/(C'*Q0*C))*C');
    Model = [Model, G0 + (Gdif *(q-q0)) <= t];
    ss = optimize(Model,t);
    q0 = round(value(q));
    Q0 = diag(q0);
    Qt = sum(q0);
    t0 = round(value(t));
    display(trace(C/(C'*Q0*C)*C'));
    ch = abs(t0 - trace(C/(C'*Q0*C)*C'));
end `

This code is not complete, because I am not calculating values for Ws. In the paper, they have not tried calculating Ws so I don't know. 

I have now matched all dimensions of all matrices and vectors for the equations to make sense. The lines with eq1 and eq2 are kind of like constraints, but I am not sure if I have to include them in Model. Are there any ideas regarding all this?

**Reply by johanlofberg** [reply]
Once again, put the code in code blocks. Your code cannot be copy-pasted to run otherwise.

Copy-paste the code to the text field. Mark it, and select code-block (<>). 

It should turn grey and clearly stand out as code in the preview, otherwise you failed.

**Reply by ArjunM21** [reply]
Okay I am sorry, but I did not know I wasn't doing it right. The errors I am getting constantly are "First argument (F) should be a constraint object." for line 43 or "The first argument should be an sdpvar object or string." for line 40, depending on how I define the constraints part.

```
clc;
clear all;
close all;
yalmip('clear');
% C = [1 0 0 ; 1 1 -1; 1 1 -1; 1 1 -1; 0 1 0; 1 0 -1; 0 1 -1; 0 0 1]; %1,5,8
C = [1 0 0 ; 0 1 0; 0 1 0; 0 1 0; 0 0 1; 0 1 -1;-1 1 0; 1 -1 1]; %1,2,5

sdpvar t;
q = binvar(8,1);
Q = diag(q);
q0 = [0;0;0;0;0;1;1;1];
Q0 = diag(q0);

syms W W1 W2 W3 W4 W5 W6 W7 W8
W = [W1 W2 W3 W4 W5 W6 W7 W8];
W1 = intvar(3,3);
W2 = intvar(3,3);
W3 = intvar(3,3);
W4 = intvar(3,3);
W5 = intvar(3,3);
W6 = intvar(3,3);
W7 = intvar(3,3);
W8 = intvar(3,3);
a = sum(W*q);
eq1 = sum(W*q) == C'*Q*C;
eq2 = sum(W) == C'*C;

Constraints = [];
Constraints = [sum(q) == 3]; 
ch = 100;
tol = 1.4e1;

Gdif = [];
for i = 1:8
    Gdif = [Gdif, -trace(inv(a)*value(W(i))*inv(a)*(C'*C))];
end

while(ch > tol)
    G0 = trace((C/(C'*Q0*C))*C');
    Constraints = [Constraints, G0 + (Gdif *(q-q0)) <= t];
%     Model = [Model, eq1, eq2];
%     options = sdpsettings('solver','mosek','verbose',2);
    ss = optimize(Constraints, t);
    q0 = round(value(q));
    Q0 = diag(q0);
    Qt = sum(q0);
    t0 = round(value(t));
    display(trace(C/(C'*Q0*C)*C'));
    ch = abs(t0 - trace(C/(C'*Q0*C)*C'));
end 
```

**Reply by johanlofberg** [reply]
What do you expect to happen when you mix sdpvar objects and syms? Of course it crashes

```
K>> class(Gdif)

ans =

    'sym'
```

**Reply by ArjunM21** [reply]
Yeah i had a feeling it had something to do with syms and sdpvar. But how else am I supposed to define Ws and q instead? I have to define all of this kinda like this for it to match the actual papaer, but I know it might be wrong. Kindly tell me a solution if you can understand the paper or my code.

**Reply by johanlofberg** [reply]
I have no idea why you would want to involve syms at all. What is it you cannot do using sdpvars. The code is massively confused. It begins by defining a set of sym scalars to create a sym vector W, then the same name is used to define 3x3 matrices instead. It writes sum(W * q) == C' * Q * C which says a scalar should be equal to a matrix (sum is redundant, and that line just runs by luck, there is no reason that syms * sdpvar should work), and then sum(W) == C' * C which says scalar equal matrix (this times completely disconnected from YALMIP as there are do sdpvars there). It looks like you in some kind of confused state simply try to create W1 * q(1)+W2 * q(2)+...

Also, the code makes no sense as you do value(W(i)) before you even solve a problem. A decision variable cannot have a value unless some optimization problem has been solved, or it explicitly has been assigned a value using assign. However, at that point W is not an sdpvar but a vector syms and value(syms) just happens to return the variable it self (i.e. by luck it doesn't crash)

**Reply by ArjunM21** [reply]
Hmm I get it. I am pretty new to this, so yeah. Let me explain what it is that I want.

I want to make a vector W consisting of 8 3x3 matrices W1, W2 up to W8. I need to be able to multiply W1 with q1, W2 with q2 and so on. Since q1, q2 etc are binary variables, multiplication of Wi and qi will be similar to adding multiple 3x3 matrices. (kind of like W1 + W2 - W3 etc). This should be equal to C'*Q*C. That is what I wanted with eq1. Now for eq2, another condition that is present is sum of W1, W2 etc should be equal to C'*C, which is a 3x3 matrix. Dimension-wise, they all match. 

Once I get this dome properly, I need to be able to calculate Gdif which is a 1x8 vector, that has 8 values corresponding to W1, W2 etc. I now believe it should go into the loop cus of your explanation. Also, I now realize that we need to get values for W1, W2 etc as well in order to solve this properly, but I am unsure how to do it. I now feel that I should give eq1 and eq2 as constraints along with my original 2 constraints and solve for q and W simultaneously while minimizing t. 

My main question is how to define the vector W consisting of 8 3x3 matrices? That is the reason I tried using syms for this.

Kindly try to make sense of all this from what I've written and feel free to ask me whatever more info you might need.

**Reply by johanlofberg** [reply]
You simply have to understand how that algebra work first without even involving yalmip. To create the matrix 

`[W1*q(1) W2*q(2) ...`

you will have to think of what that is in linear algebra. It is 

`[W1 W2 ...] .*[q(1)*I q(2)*I...]`

If you have created W by using cell format (I presume you've read the basic tutorial), the most compact code would be

`[W{:}].*kron(q',eye(3))`

and the sum 

`W1*q(1) + W2*q(2)`

would by same linear algebra be

`[W{:}]*kron(q,eye(3))`

**Reply by ArjunM21** [reply]
Yeah I understand the linear algebra. But I cannot define W as a cell, cus then how do I assign 3x3 matrices to it?

**Reply by johanlofberg** [reply]
W can be a cell, read the basic tutorial on how to do it quickly, or simply place sdpvars manually in a cell

**Reply by ArjunM21** [reply]
So I did something like this. I understand the kron part but won't that make it a 8x24 matrix? According to me the W is a 1x8 cell right? Is this how I should go about this?

```
W1 = cell(3);
W2 = cell(3);
W3 = cell(3);
W4 = cell(3);
W5 = cell(3);
W6 = cell(3);
W7 = cell(3);
W8 = cell(3);
W = {W1, W2, W3, W4, W5, W6, W7, W8};
AA = W*q;
```

**Reply by johanlofberg** [reply]
Did you really go back and read the basic tutorial?

**Reply by ArjunM21** [reply]
Are you talking about a specific tutorial? If so, kindly share it with me

**Reply by johanlofberg** [reply]
https://yalmip.github.io/tutorial/basics/

**Reply by ArjunM21** [reply]
So something like this? It ran without error, so I guess its better than the last one. I can use Sum1 ans Sum2 as constraints. Now, how do I use this to define Gdif inside the loop? Should I use an initial value for W to run the loop once?


```
W1 = sdpvar(3);
W2 = sdpvar(3);
W3 = sdpvar(3);
W4 = sdpvar(3);
W5 = sdpvar(3);
W6 = sdpvar(3);
W7 = sdpvar(3);
W8 = sdpvar(3);
W = {W1, W2, W3, W4, W5, W6, W7, W8};
Sum1 = [W{:}]*kron(q,eye(3));
Sum2 = [W{:}]*kron(ones(8,1),eye(3));
```

**Reply by johanlofberg** [reply]
More compact code using the cell format in the tutorial

`W = sdpvar(repmat(3,1,8))`

First sum will be quick code for W{1} * q(1) + .... while the second is W{1} +W{2} + ... so if that is what you want

**Reply by ArjunM21** [reply]
Yeah this is right. But now I wanna code Gdif. Do you have any idea regarding that? I guess it has to be in the loop, but I'm not sure.

**Reply by ArjunM21** [reply]
What I mean is in the code given below, I want to enter W1, W2, W3 etc in the place of the curly bracket with 4 stars. But I don't know how to do it, as I am unable to extract just that cus right now W(1) will give W1 as a cell and cell2mat doesn't work.

```
Gdif = [];
for i = 1:8
    Gdif = [Gdif, -trace(inv(Sum1)*{****}*inv(Sum1)*(C'*C))];
end
```

**Reply by johanlofberg** [reply]
Since the variables are placed in a cell, you have to extract them using cell indexing, W{i}

**Reply by ArjunM21** [reply]
Yeah I tried them but came to realize that YALMIP doesn't support inverses. So the inv(Sum1) part won't work, since Sum1 is an sdpvar as well. And I read your other replies saying there isn't a specific way to solve this unless we use X*Y==eye(n) as a constraint and try. 

Thank you for your help. If you have any more ideas, kindly share them with me.

**Reply by johanlofberg** [reply]
Isn't that a linearization strategy, i.e. the inverse is on the current solution, not the decision variable. Otherwise, it makes no sense as your linearization is not linear...

**Comment by ArjunM21** [comment]
Hey Johan.
Thank you for your help previously. It has given me a few ideas and I was able to complete the code. Please see it below. Now when I run the program, it is doing what it is supposed to do, but the results are very much off. Can you tell me what is happening? For now, all dimensions match and the program is running kinda okay. I need to add one more constraint for the Sum1 case, which when I did, resluted in an infinite loop somehow I guess. If this code makes sense to you, could you explain why I am getting the result I get when I run it?
```

clc;
clear all;
close all;
yalmip('clear');
% C = [1 0 0 ; 1 1 -1; 1 1 -1; 1 1 -1; 0 1 0; 1 0 -1; 0 1 -1; 0 0 1]; %1,5,8
C = [1 0 0 ; 0 1 0; 0 1 0; 0 1 0; 0 0 1; 0 1 -1;-1 1 0; 1 -1 1]; %1,2,5

sdpvar t;
q = binvar(8,1);
Q = diag(q);
q0 = [0;1;0;0;1;0;1;0];
Q0 = diag(q0);

W1 = sdpvar(3);
W2 = sdpvar(3);
W3 = sdpvar(3);
W4 = sdpvar(3);
W5 = sdpvar(3);
W6 = sdpvar(3);
W7 = sdpvar(3);
W8 = sdpvar(3);
W = {W1, W2, W3, W4, W5, W6, W7, W8};
Sum1 = [W{:}]*kron(q,eye(3));
Sum2 = [W{:}]*kron(ones(8,1),eye(3));

Constraints = [];
Constraints = [sum(q) == 3]; 
ch = 100;
tol = 1e-2;

Gdif = [];

while(ch > tol)
    G0 = trace((C/(C'*Q0*C))*C');
    for i = 1:8
        Gdif = [Gdif, -trace(inv(C'*Q0*C)*(W{i})*inv(C'*Q0*C)*(C'*C))];
    end 
    Constraints = [Constraints, G0 + (Gdif *(q-q0)) <= t];
    Constraints = [Constraints, Sum2 == C'*C];
    ss = optimize(Constraints, t);
    q0 = round(value(q));
    Q0 = diag(q0);
    Qt = sum(q0);
    t0 = round(value(t)); 
    display(trace(C/(C'*Q0*C)*C'));
    disp('a');
    ch = abs(t0 - trace(C/(C'*Q0*C)*C'));
    Gdif = [];
end

**Reply by johanlofberg** [reply]
Surely it is supposed to be the current value of W being used in Gdif, not the decision variable. As it is now, your "linearization" is not linear as you multiply W and q, thus bilinear and extremely hard. If it is really intended to be like this, well then good luck as you have a very very hard problem

**Reply by johanlofberg** [reply]
and why not W = sdpvar(3*ones(1,8)) since you read the basic tutorial where this was discussed

**Reply by ArjunM21** [reply]
Hey thanks.
No it;'s not intended to be like this. I now realize that it is bilinear, which is difficult to solve. How do I make use of the current value of W instead?
Also, I Hadn't seen that usage of sdvpar, sorry. But thanks, let me try it out. Also I found out another way to do the same, so maybe I will try that as well. Thank you.

**Reply by johanlofberg** [reply]
Just like you do with the other variables where you initialize to some value and then iteratively extract the last solution, i.e. effectively use value(W{i}) in the loop once initialized

**Reply by ArjunM21** [reply]
Hey Johan.
Thank you so much for being so helpful. I am trying out something else before proceeding further. Please see the code below. I want to know why the value of q0 (or q) doesn't reduce to give the optimal value? As of now, it returns the initial value of q0 that I defined. Doesn't that mean it isn't optimizing? I think the constraints that I have given aren't considered because the value of Q isn't being updated. Can you help me understand what's wrong here?

```
clc;
clear all;
close all;
yalmip('clear');
C = [1 0 0 ; 1 1 -1; 1 1 -1; 1 1 -1; 0 1 0; 1 0 -1; 0 1 -1; 0 0 1]; %1,5,8
% C = [1 0 0 ; 0 1 0; 0 1 0; 0 1 0; 0 0 1; 0 1 -1;-1 1 0; 1 -1 1]; %1,2,5

sdpvar t;
q = binvar(8,1);
Q = diag(q);
q0 = [1;0;0;0;0;1;1;0];
Q0 = diag(q0);

W1 = [1 0 0;0 0 0;0 0 0];
W2 = [1 1 -1;1 1 -1;-1 -1 1];
W3 = [1 1 -1;1 1 -1;-1 -1 1];
W4 = [1 1 -1;1 1 -1;-1 -1 1];
W5 = [0 0 0;0 1 0;0 0 0];
W6 = [1 0 -1;0 0 0;-1 0 1];
W7 = [0 0 0;0 1 -1;0 -1 1];
W8 = [0 0 0;0 0 0;0 0 1];
W = {W1, W2, W3, W4, W5, W6, W7, W8};
Sum1 = [W{:}]*kron(q,eye(3));
Sum2 = [W{:}]*kron(ones(8,1),eye(3));

Constraints = [];
Constraints = [sum(q) == 3]; 
ch = 100;
tol = 1e-2;

Gdif = [];

while(ch > tol)
    G0 = trace((C/(C'*Q0*C))*C');
    for i = 1:8
        Gdif = [Gdif, -trace(inv(C'*Q0*C)*value(W{i})*inv(C'*Q0*C)*(C'*C))];
    end 
    Constraints = [Constraints, G0 + (Gdif *(q-q0)) <= t];
    Constraints = [Constraints, Sum2 == C'*C];
    Constraints = [Constraints, Sum1 == C'*Q0*C];
    ss = optimize(Constraints, t);
    q0 = round(value(q));
    Q0 = diag(q0);
    Qt = sum(q0);
    t0 = value(t); 
    
    disp('a');
    ch = abs(t0 - trace(C/(C'*Q0*C)*C'));
    Gdif = [];
end 
```

**Reply by johanlofberg** [reply]
This discussion forum is about YALMIP specific issues. This is a question about convergence properties of some algorithm. You are better off contacting the authors proposing that algorithm, or if it is your own, discuss with your supervisor.
